Python 3 compatibility

dev-zero opened the pull request:
The following fixers from [Python Modernize](https://python-modernize.readthedocs.io/en/latest/) were run and their changes completely reviewed, with each fixer being at least one commit. The following list is a mirror of the python-modernize hook configuration in `.pre-commit-config.yaml`.

- [x] lib2to3.fixes.fix_apply  (apply)
- [x] lib2to3.fixes.fix_except  (except)
- [x] lib2to3.fixes.fix_exec  (exec)
- [x] lib2to3.fixes.fix_execfile  (execfile)
- [x] lib2to3.fixes.fix_exitfunc  (exitfunc)
- [x] lib2to3.fixes.fix_funcattrs  (funcattrs)
- [x] lib2to3.fixes.fix_has_key  (has_key) **false positives**
- [ ] lib2to3.fixes.fix_idioms  (idioms) *optional*
- [x] lib2to3.fixes.fix_long  (long) **false positives**
- [x] lib2to3.fixes.fix_methodattrs  (methodattrs)
- [x] lib2to3.fixes.fix_ne  (ne)
- [x] lib2to3.fixes.fix_numliterals  (numliterals)
- [x] lib2to3.fixes.fix_operator  (operator)
- [x] lib2to3.fixes.fix_paren  (paren)
- [x] lib2to3.fixes.fix_reduce  (reduce)
- [x] lib2to3.fixes.fix_renames  (renames)
- [x] lib2to3.fixes.fix_repr  (repr)
- [ ] lib2to3.fixes.fix_set_literal  (set_literal) *optional*
- [x] lib2to3.fixes.fix_standarderror  (standarderror)
- [x] lib2to3.fixes.fix_sys_exc  (sys_exc)
- [x] lib2to3.fixes.fix_throw  (throw)
- [x] lib2to3.fixes.fix_tuple_params  (tuple_params)
- [x] lib2to3.fixes.fix_types  (types)
- [ ] lib2to3.fixes.fix_ws_comma  (ws_comma) *optional*
- [x] lib2to3.fixes.fix_xreadlines  (xreadlines)
- [x] libmodernize.fixes.fix_basestring  (basestring)
- [x] libmodernize.fixes.fix_classic_division  (classic_division)
- [x] libmodernize.fixes.fix_dict_six  (dict_six)
- [x] libmodernize.fixes.fix_file  (file)
- [x] libmodernize.fixes.fix_filter  (filter)
- [x] libmodernize.fixes.fix_import  (import)
- [x] libmodernize.fixes.fix_imports_six  (imports_six)
- [x] libmodernize.fixes.fix_input_six  (input_six)
- [x] libmodernize.fixes.fix_int_long_tuple  (int_long_tuple)
- [x] libmodernize.fixes.fix_itertools_imports_six  (itertools_imports_six)
- [x] libmodernize.fixes.fix_itertools_six  (itertools_six)
- [x] libmodernize.fixes.fix_map  (map)
- [x] libmodernize.fixes.fix_metaclass  (metaclass)
- [x] libmodernize.fixes.fix_next  (next)
- [ ] libmodernize.fixes.fix_open  (open)
- [x] libmodernize.fixes.fix_print  (print)
- [x] libmodernize.fixes.fix_raise  (raise)
- [x] libmodernize.fixes.fix_raise_six  (raise_six)
- [x] libmodernize.fixes.fix_unichr  (unichr)
- [x] libmodernize.fixes.fix_unicode  (unicode)
- [x] libmodernize.fixes.fix_unicode_future  (unicode_future)
- [x] libmodernize.fixes.fix_unicode_type  (unicode_type)
- [x] libmodernize.fixes.fix_urllib_six  (urllib_six)
- [x] libmodernize.fixes.fix_xrange_six  (xrange_six)
- [x] libmodernize.fixes.fix_zip  (zip)

Entries marked with *optional* are fixers which are not required to run to make the codebase Python 3 compatible. Such marked with **false positives** have been run, their changes applied if correct, but they also generate undesired changes or even break correct code.

The current state is that all files can be compiled with Python 3 and `pip install` works with Python 3.

Why to use Python Modernize instead of Python Futurize: Python-Futurize seems to be dead (last release in 2016), while there are still commits to both Modernize and Six.

ltalirz commented:
@dev-zero just letting you know that we are currently in the process of migrating the command line to click - with the upside that, finally, this part will be tested properly as well (see the `verdi` branch).
For the python3 migration it could be advantageous to wait until this merge has happend (just saying,  not to stop you from proceeding if you like to).

dev-zero commented:
@ltalirz thanks for the heads-up, I already noticed, hence my comment `Some of this PR probably clashes with the click-migration PRs.`. So, most of the changes in `aiida/cmdline/` can then probably be dropped, while there a lot of other places which still need to be touched.

giovannipizzi commented:
@greschd do you have any comment on the comment by @dev-zero ? 
Would it be possible to enforce to always use UTF8 for strings, both when storing and caching?

giovannipizzi commented:
@dev-zero: we discussed this today at the AiiDA meeting.
The suggestion would be
- wait for the final 'verdi' merge (should happen this week if all goes week)
- continue what you are doing
- as soon as you have some of the important changes done, and some kind of tests can be added to travis (e.g. running the tests on python3, or at least some precommits to avoid people introduce changes that are not python3-compatible) we could already merge the PR, keeping the issue open, and create a new branch/PR to finalise the work. This make sure that changes happening in the meantime don't revert all the work you've done.
(Of course, this assumes that (at least under python2) things keep working properly - if this is not the case, then this suggestion does not apply).

Also, for this preliminary PR, it would be great to have a (very brief) summary on how to do things to be both py2 and py3 compatible, e.g. when creating new files-I imagine things like
- use this library and not this one (e.g. between futurize, modernize, six, ...)
- start all files with `from __future__ import ...`

This will be useful already now to keep compatibility with both, and also for the plugin developers to follow the same guidelines

giovannipizzi commented:
I think the equivalence you mention might be correct, even if any feedback is welcome.
A question: what should be considered unicode and what bytestring? Should we always use utf-8 as encoding, and then hash the corresponding decoded bytestring, for anything stored in the DB? (and in this case, do we need to be careful about the encoding of the DB (postgres) or not?)
Instead, for files in the repository, should we always use directly the bytestring?

DropD commented:
By the way, verdi-click is now merged and there are no sweeping changes to be expected from that direction anymore.

greschd commented:
I would distinguish bytestring and unicode explicitly, by giving them different salts (e.g. 'b' and 'u'). For the unicode strings, we just need to pick an encoding because only then can it be hashed -- it makes sense to use utf-8, but actually would work with any encoding.

For the things stored in the DB I'm not sure if it's possible to accidentally change the type when doing the Python - DB - Python roundtrip. If so, that might create issues when things hash differently after being stored.

For files in the repository, we just use the filename + (byte) content -- that should not be affected by py2 / py3, except when there's a bug in how the file is written.

dev-zero commented:
I agree with @greschd . For testing I suggest we add a test with the different objects and their respective hashes with Python 2.7 generated on a `en_US.utf-8` system to make sure we don't have any regressions compared with that reference.

dev-zero commented:
I rolled back most of the commits for now (they are still in my branch [py3-compat-wip](https://github.com/dev-zero/aiida_core/tree/py3-compat-wip)). I'm going to re-add them here once I got tests up and running for each of the changes to avoid any regressions.
The first one going to be dependencies and the `setup.py`.

dev-zero commented:
To make Yapf happy, I added compat fix for the calls to `exec` (which the automated fixer got wrong in one case btw). Running `python-modernize --fix=exec` is therefore now part of the pre-commit hook and works with both Python 2 and 3. Additional checks can be added there after fixing them in the code.
Since the Pattern package was actually not used, I removed it for now. Version 3.6 of the package is actually Python 3 compatible, but pulls in MySQL as a dependency. If we need that package in the future, someone should probably talk to the Pattern devs to remove that dep again.

dev-zero commented:
Btw, I intentionally did not add `modernize` to `setup_requirements.py` since `pre-commit` downloads a more recent version by itself anyway when installing the hooks.

dev-zero commented:
For those interested, the easiest way to help is by:

1. uncommenting a commented out fixer in `.pre-commit-config.yaml` (works for both Python 2 and 3)
2. run `pre-commit run -a python-modernize`
3. check the changes
4. commit (using Python 2 since pylint still fails with Python 3)

dev-zero commented:
@DropD since you are working on the quicksetup you may want to pull the related changes from 17d8f9a for it (replaced all `print` statements with `echo.echo`).

dev-zero commented:
Maybe as a note for the review process: replacing things like `filter`/`map`/`zip` by their iterator-equivalents is not only about making sure whether in that place an `iterator` is an acceptable type (so, whether there is an index operation `[0]` on it), but when used in `for`-loops it is also important that the data used for `filter`/`map`/`zip` is not modified: while this would have worked in Python 2 since first a list with copies was built and then iterated, this does not work with iterators.

dev-zero commented:
ok, I have almost everything in place in https://github.com/dev-zero/aiida_core/tree/py3-compat-next to actually start testing. The only roadblock is currently the PyCifRW update, which seems to be an upstream bug I reported here: https://bitbucket.org/jamesrhester/pycifrw/issues/14/pycifrw-44-fails-with-python-27-and-long

dev-zero commented:
Ok, after those last commits we can start with running the unittests (actually already started on my py3-compat-next branch).

One thing I already encountered when trying to build the documentation: the currently used version of Django is not Python 3.4+ compatible, see for example https://github.com/django/django/commit/b07aa52e8a8e4c7fdc7265f75ce2e7992e657ae9
Is it an option to just not support Django with Python 3?

giovannipizzi commented:
Mmm, I fear we need to keep django still a bit... do you know what is the first version of django that might work? So we can check how much django charged and how hard it is to upgrade (most of the changes are often on the web part that we don't use, we just need to be careful to changes to the db access, like transactions etc.)

dev-zero commented:
@giovannipizzi at least Django 1.8a1 to include that specific fix. Version 1.8.19 seems to be the latest in that LTS branch whose support unfortunately just has expired (April 1st). So, if possible we should upgrade to Django 1.11, which is still going to be supported until 2020 and which is also the last version of Django supporting Python 2.7.

dev-zero commented:
The upgrade to Django 1.8.19 requires a small migration for the `last_login` and `email` columns in `dbuser`, after that the tests seem to work, see commit 768d09e.

giovannipizzi commented:
Thanks Tiziano!
Just one comment/question on the migration - we've been already wondering (and don't have a solution) if it's possible to have a non-linear history in the migrations - there is the risk that in develop we do another migration, that would end up having the same ID/integer position, unless this is merged soon. Do you have any solutions/ideas for this? Or would it be maybe possible to do just the migration in a different PR, if this is a solution/workaround?

dev-zero commented:
After some more digging, it seems there is at least one fundamental issue with the hashing: it relies on the CPython 2 comparison rules for comparison between different datatypes (it is even implementation dependent!):

```python
@make_hash.register(abc.Set)
def _(object_to_hash, **kwargs):
    hashes = tuple([
            make_hash(x, **kwargs)
            for x
            in sorted(object_to_hash)   # <=== here
        ])
    return make_hash_with_type('S', ",".join(hashes))

@make_hash.register(abc.Mapping)
def _(mapping, **kwargs):
    hashed_dictionary = {
        k: make_hash(v, **kwargs)
        for k,v
        in mapping.items()
    }
    return make_hash_with_type(
        'D',
        make_hash(sorted(hashed_dictionary.items()), **kwargs)  # <=== and here
    )
```

The example in the `make_hash` code explicitly includes sets and lists with mixed datatypes (as do real-world examples), namely `int` and `str`, which are not comparable anymore in Python 3. There are ways to emulate the Python 2 behavior in Python 3, but is that really the way to go? Also because that solution requires additional sorting and ordering which is likely to be slower.

The alternative would be to sort the hashes of the items before combining them, breaking backward compatibility with Python 2 generated hashes.

dev-zero commented:
@giovannipizzi sure, I will do a separate PR just for the Django 1.8 upgrade.

While there will be more commits coming, I would consider this PR ready for review (and merging):

* The commits are all standalone in the sense that I squashed later fixes for changes introduced by those.
* Every commit is tested via CI.
* While there will possibly be some changes which could be merged into one of the already present commits, they will come slower, since I'm now in the manual testing&fixing feedback loop. So, they can go in as commits of their own.
* I added longer commit messages with explanations where there was more to change than what was proposed by the python-modernize fixer.
* Everything is now at least Python 3 syntax compatible and the documentation generation and the pre-commit hooks work on Python 3.

sphuber commented:
@dev-zero I merged your PR to update django to support python3. You can now rebase these commits if you want to include the fix. After that, once the final few steps are done, I think we can merge this. Thanks a lot

dev-zero commented:
@sphuber cool, thanks! many of the changes up to b5e6bb5 are covered by running python-modernize and/or pylint+sphinx (see the commits which usually also include a change to `.pre-commit-config.yaml` to include the respective fixer). But I'm already a step ahead: Today I finished the Django backend tests except for the following (see also commit 16c46ba):

* issue #1916 
* everything PyCifRW-related, due to [not having one version which is Python 2 and 3 compatible](https://bitbucket.org/jamesrhester/pycifrw/issues/14/pycifrw-44-fails-with-python-27-and-long). Unfortunately this also includes the RESTApi test suite for now.
* all Tcod, Pcod, Mpod import/export: they seem rather thoroughly broken wrt byte/str handling
* one pymatgen test, again a str/byte problem if I remember correctly
* hashing for mixed-key dics (`{1:1, '1':1}`), see earlier messages

The SQLAlchemy backend needs some more love, as you can see in the [Travis logs of my branch](https://travis-ci.org/dev-zero/aiida_core/jobs/422194025).

... unfortunately it seems that the TransportQueue also needs some more work (got skipped on my test machine for some reason).

Will give it another shot tomorrow...

dev-zero commented:
ok, SQLAlchemy tests seem to be fixed as well, just have to figure out the errors I get on Travis but not locally.
The fix in b03dd46 might be needed in other places where a custom `__getattr__` was added, haven't checked all of them yet, though. Plus, I still don't know why it did not show up on Python 2. Maybe the  the implementation of `copy.copy` changed, because we were using new-style classes which should give the same behavior wrt to `__getattr__` and attribute resolution.

sphuber commented:
@dev-zero yeah that is something @nmounet  added, but I was tangentially involved, so I will have a look and put out a fix

dev-zero commented:
ok, the Transport problems seem to be a bit more involved, especially since I can't reproduce them in my Docker environment for debugging. Disabled the respective tests as well.
Should I open bugs for the remaining Python 3 incompatibilities?

dev-zero commented:
The issues with Transport should be fixed with 4960ac8. I took the liberty to clean up the pre-commit configuration, this can go in afterwards (please note that GitHub orders the commits by their date while I re-sorted them to enable tests after the fixes)

sphuber commented:
@dev-zero we just released v0.12.2 and I need to merge `master` back into `develop`. I was now pondering how this will influence this PR and what the best order would be. What do you think would be best. I merge `master` into `develop` and you rebase, or we wait with the merge after this is merged. For the latter, do you have an idea how much longer we need to get this PR merged and second, would the rebase after merge of master cause you a lot of trouble you think?

dev-zero commented:
@sphuber even though I would prefer to have this PR merged first since that would mean less work for me, it is probably better to first merge the master and then rebase (also because I have a test environment for Python 3).
Since the issues with PyCifRW, the COD classes and the Transport take more time, I think it would make sense to merge this PR right afterwards. If you do not have time to fix #1916 anytime soon, I can also disable this test for now to make Travis happy.

dev-zero commented:
btw, I would probably not squash this commit when merging, but keep the commits as such: if there is a problem with a change introduced by one of this commits in one place, it might affect other parts with the same change as well. Having the specific commit makes it easier to identify them.

sphuber commented:
Totally agree that this PR should not be squashed, that was already the plan. I started working on #1916 but besides the faulty use of generator variables outside the comprehension, it seems like the logic in principle is weird and broken. I tried to fix it by implementing what I think the function was supposed to do, but that breaks the tests. This could be just be because now the correct logic is happening and we should just adapt the test expectations, but for that I need to confirm with @nmounet who wrote the function. This might take a bit longer therefore.

dev-zero commented:
@sphuber ok, rebased my branch `py3-compat-next` and I'm testing/reviewing again.

I get the following test failures for the TransportQueue on Travis:

```
======================================================================
ERROR: test_get_transport_interleaved (aiida.backends.tests.work.test_transport.TestTransportQueue)
Test interleaved calls to get the same transport
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dev-zero/aiida_core/aiida/backends/tests/work/test_transport.py", line 69, in test_get_transport_interleaved
    loop.run_sync(lambda: [interleaved(self.authinfo), interleaved(self.authinfo)])
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/tornado/ioloop.py", line 456, in run_sync
    if not future_cell[0].done():
AttributeError: 'NoneType' object has no attribute 'done'
======================================================================
ERROR: test_get_transport_nested (aiida.backends.tests.work.test_transport.TestTransportQueue)
Test nesting calls to get the same transport
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dev-zero/aiida_core/aiida/backends/tests/work/test_transport.py", line 57, in test_get_transport_nested
    loop.run_sync(lambda: nested(transport_queue, self.authinfo))
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/tornado/ioloop.py", line 457, in run_sync
    raise TimeoutError('Operation timed out after %s seconds' % timeout)
tornado.ioloop.TimeoutError: Operation timed out after None seconds
[...]
```
I am unable to reproduce it in my Docker for now, but that might be because I don't have the full torquessh machine setup inside.

Yes, I will open new issues for PyCifRW and the importers. For the importers/exporters: Unfortunately there seem to be general problems hidden in there (which are also affecting Python 2). Examples:
* missing default encoding when writing/reading files which may break under different system encoding and/or when changing the default system encoding between save&load. Also, some tests start failing when running with Japanese or Chinese locale.
* including file contents as strings in JSON files without prior base64 encoding which could introduce subtle bugs and differences (for example in hashing or due to double encoding)

dev-zero commented:
I did some more reshuffling of the commits but now it's ready. Up to a91b1c0a747d34b54af33a20cacf62565cfa6283 it's all strictly Python 3 related. Everything after it could go into one or more separate PRs. GitHub seems to list commits still in the wrong order.

Do we need to write down some developer guidelines? Many things will get corrected by python-modernize, but some are not. Like the fact that the `dict.items()`, `.values()`, `.keys()` in Python 3 behave more like Python 2's `.iteritems()`. So if people need those functions to return a full list, they have to wrap it in a `list(d.items())`.

If you want to see the Python 3 test failures, simply `git revert 123033b` in a separate branch.

ltalirz commented:
Fantastic, thanks a lot @dev-zero !
